dipy/io/stateful_tractogram.py

```diff
@@ -0,0 +1,746 @@
+            self._rasmm_to_voxmm()
+
+    def to_rasmm(self):
+        """ Safe function to transform streamlines and update state """
+        if self._space == Space.VOX:
+            self._vox_to_rasmm()
+        elif self._space == Space.VOXMM:
+            self._voxmm_to_rasmm()
+
+    def to_center(self):
+        """ Safe function to shift streamlines so the center of voxel is
+        the origin """
+        if self._shifted_origin:
+            self._shift_voxel_origin()
+
+    def to_corner(self):
+        """ Safe function to shift streamlines so the corner of voxel is
+        the origin """
+        if not self._shifted_origin:
+            self._shift_voxel_origin()
+
+    def compute_bounding_box(self):
+        """ Compute the bounding box of the streamlines in their current state
+
+        Returns
+        -------
+        output : ndarray
+            8 corners of the XYZ aligned box, all zeros if no streamlines
+        """
+        if self._tractogram.streamlines.data.size > 0:
+            bbox_min = np.min(self._tractogram.streamlines.data, axis=0)
+            bbox_max = np.max(self._tractogram.streamlines.data, axis=0)
+
+            return np.asarray(list(product(*zip(bbox_min, bbox_max))))
+
+        return np.zeros((8, 3))
+
+    def is_bbox_in_vox_valid(self):
+        """ Verify that the bounding box is valid in voxel space
+        Will transform the streamlines for OBB, slow for big tractogram
+
+        Returns
+        -------
+        output : bool
+            Are the streamlines within the volume of the associated reference
+        """
+        old_space = deepcopy(self.get_current_space())
+        old_shift = deepcopy(self.get_current_shift())
+
+        # Do to rotation, equivalent of a OBB must be done
+        self.to_vox()
+        self.to_corner()
+        bbox_corners = deepcopy(self.compute_bounding_box())
+
+        is_valid = True
+        if np.any(bbox_corners < 0):
+            logging.error('Voxel space values lower than 0.0')
+            logging.debug(bbox_corners)
+            is_valid = False
+
+        if np.any(bbox_corners[:, 0] > self._dimensions[0]) or \
+                np.any(bbox_corners[:, 1] > self._dimensions[1]) or \
+                np.any(bbox_corners[:, 2] > self._dimensions[2]):
+            logging.error('Voxel space values higher than dimensions')
+            logging.debug(bbox_corners)
+            is_valid = False
+
+        if old_space == Space.RASMM:
+            self.to_rasmm()
+        elif old_space == Space.VOXMM:
+            self.to_voxmm()
+
+        if not old_shift:
+            self.to_center()
+
+        return is_valid
+
+    def _get_streamline_count(self):
+        """ Safe getter for the number of streamlines """
+        return len(self._tractogram.streamlines._offsets)
+
+    def _get_point_count(self):
+        """ Safe getter for the number of streamlines """
+        return len(self._tractogram.streamlines.data)
```

Comment:
If the `streamlines` object (i.e. an `ArraySequence`) is the result of a slicing or advanced indexing operation (and no `.copy()` was called on it) taking the `len` of `.data)` would be wrong.
```suggestion
        return self._tractogram.streamlines.total_nb_rows
```
Under the hood (see https://github.com/nipy/nibabel/blob/master/nibabel/streamlines/array_sequence.py#L115), it sums the lengths of the remaining sliced/indexed streamlines.